peps/pep-0765.rst

```diff
@@ -149,11 +151,169 @@ How to Teach This
+appear in very popular libraries, including a cloud security application.
+For those so inclined, it should not be hard to replicate my analysis, as I
+provided links to the scripts I used in the Method section.
+
+The projects examined contained a total of 120,964,221 lines of Python code,
+and among them the script found 203 instances of control flow instructions in a
+``finally`` block.  Most were ``return``, a handful were ``break``, and none were
+``continue``. Of these:
+
+- 46 are correct, and appear in tests that target this pattern as a feature (e.g.,
+  tests for linters that detect it).
+- 8 seem like they could be correct - either intentionally swallowing exceptions
+  or appearing where an active exception cannot occur. Despite being correct, it is
+  not hard to rewrite them to avoid the bad pattern, and it would make the code
+  clearer: deliberately swallowing exceptions can be more explicitly done with
+  ``except BaseException:``, and ``return`` which doesn't swallow exceptions can be
+  moved after the ``finally`` block.
+- 149 were clearly incorrect, and can lead to unintended swallowing of exceptions.
+  These are analyzed in the next section.
+
+**The Error Cases**
+
+Many of the error cases followed this pattern:
+
+.. code-block::
+    :class: bad
+
+    try:
+        ...
+    except SomeSpecificError:
+        ...
+    except Exception:
+        logger.log(...)
+    finally:
+        return some_value
+
+Code like this is obviously incorrect because it deliberately logs and swallows
+``Exception`` subclasses, while silently swallowing ``BaseExceptions``. The intention
+here is either to allow ``BaseExceptions`` to propagate on, or (if the author is
+unaware of the ``BaseException`` issue), to log and swallow all exceptions. However,
+even if the ``except Exception`` was changed to ``except BaseException``, this code
+would still have the problem that the ``finally`` block swallows all exceptions
+raised from within the ``except`` block, and this is probably not the intention
+(if it is, that can be made explicit with another ``try``-``except BaseException``).
+
+Another variation on the issue found in real code looks like this:
+
+.. code-block::
+    :class: bad
+
+        try:
+            ...
+        except:
+            return NotImplemented
+        finally:
+            return some_value
+
+Here the intention seems to be to return ``NotImplemented`` when an exception is
+raised, but the ``return`` in the ``finally`` block would override the one in the
+``except`` block.
+
+> [!NOTE]
+> Following the
+> `discussion <https://discuss.python.org/t/an-analysis-of-return-in-finally-in-the-wild/70633/15>`__,
+> I repeated the analysis on a random selection of PyPI packages (to
+> analyze code written by *average* programmers). The sample contained
+> in total 77,398,892 lines of code with 316 instances of ``return``/``break``/``continue``
+> in ``finally``. So about 4 instances per million lines of code.
```

Comment:
This hasn't rendered properly:

https://pep-previews--4125.org.readthedocs.build/pep-0765/#results

The rst syntax is like this:

https://docutils.sourceforge.io/docs/ref/rst/directives.html#note

So something like this (I've not tested it):

```suggestion
.. note:: Following the
   `discussion <https://discuss.python.org/t/an-analysis-of-return-in-finally-in-the-wild/70633/15>`__,
   I repeated the analysis on a random selection of PyPI packages (to
   analyze code written by *average* programmers). The sample contained
   in total 77,398,892 lines of code with 316 instances of ``return``/``break``/``continue``
   in ``finally``. So about 4 instances per million lines of code.
```